File: C#/Leetcode/String/ReverseVowelsOfAString.cs
using System.Collections.Generic;
using System.Text;

namespace LeetcodeSolutions.String
{
    // Leetcode 345
    // Submission: https://leetcode.com/submissions/detail/140495752/
    // 2 pointers
    public class ReverseVowelsOfAString
    {
        //"hello", return "holle"
        //"leetcode", return "leotcede"
        //"google", return "geoglo"
        //"letcode", return "letcode"
        private static HashSet<char> vowels = new HashSet<char>() {
            'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U' };

        // Tx = O(n)
        // Sx = O(n)    // for the new reversed string
        public string ReverseVowels(string s)
        {
            if (s == null || s.Length == 0)
                return s;

            int low = 0, high = s.Length - 1;
            StringBuilder sb = new StringBuilder(s);

            while (low < high)
            {
                // Keep going right until low < high and char at index low is a valid vowel.
                while (low < high && !vowels.Contains(sb[low]))
                    low++;

                // Keep going left until low < high and char at index high is a valid vowel.
                while (low < high && !vowels.Contains(sb[high]))
                    high--;

                // case sensitive character comparision
                if (!sb[low].Equals(sb[high]))
                {
                    char temp = sb[low];
                    sb[low] = sb[high];
                    sb[high] = temp;
                }

                low++; high--;
            }

            return sb.ToString();
        }
    }
}

File: Java/TwoPointers/ReverseVowelsofAString.java
package TwoPointers;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class ReverseVowelsofAString {
    // Leetcode 305: https://leetcode.com/problems/reverse-vowels-of-a-string/description/
    // Submission Detail: https://leetcode.com/problems/reverse-vowels-of-a-string/submissions/1667687477

    //Input: s = "IceCreAm"
    //Output: "AceCreIm"
    //Explanation:
    //The vowels in s are ['I', 'e', 'e', 'A']. On reversing the vowels, s becomes "AceCreIm".

    //Input: s = "leetcode"
    //Output: "leotcede"

    static Set<Character> vowels = new HashSet<>(List.of('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'));
    public String reverseVowels(String s) {
        int left = 0, right = s.length()-1;
        StringBuilder sb = new StringBuilder(s);

        while(left < right) {
            while(left < right && !vowels.contains(s.charAt(left))) {
                left++;
            }

            while(left < right && !vowels.contains(s.charAt(right))) {
                right--;
            }

            swap(left, right, sb);
            left++;
            right--;
        }

        return sb.toString();
    }

    private void swap(int i, int j, StringBuilder sb) {
        char temp = sb.charAt(i);
        sb.setCharAt(i, sb.charAt(j));
        sb.setCharAt(j, temp);
    }
}
